File: SlidePuzzle/solve4-1.pl
#!/usr/bin/perl

use utf8;
use strict;
use warnings;

use Time::HiRes qw(gettimeofday);

#my $puzzle = &new_puzzle(3, 3, '120743586');
#my $puzzle = &new_puzzle(3, 3, '168452=30');
#my $puzzle = &new_puzzle(3, 4, '1365720A984B');
my $puzzle = &new_puzzle(3, 4, '4127=36B89A0');
#my $puzzle = &new_puzzle(3, 5, 'D42C7380915AB6E');
#my $puzzle = &new_puzzle(4, 4, '41EC86079BA532FD');
my $OPT = 1;

my $start = gettimeofday();
my $answer = &solve_puzzle($puzzle);
my $end = gettimeofday();
if (defined $answer) {
    printf "answer=%s (in %f sec)\n", $answer, ($end - $start);
} else {
    print "Not found\n";
}

sub solve_one {
    my ($w, $h, $s) = @_;
    my $puzzle = &new_puzzle($w, $h, $s);
    my $answer = &solve_puzzle($puzzle);
    undef $puzzle;
    return $answer;
}

sub new_puzzle {
    my ($w, $h, $s) = @_;
    my $puzzle = {
        w => $w,
        h => $h,
        first => $s,
        final => &get_final($s),
        queue => [$s],
        hash => { $s => '' },
    };
    return $puzzle;
}

sub solve_puzzle {
    my ($puzzle) = @_;
    my $count = 0;
    my $moval = {
        U => -$puzzle->{w},
        L => -1,
        R =>  1,
        D =>  $puzzle->{w},
    };
    my ($w, $h) = ($puzzle->{w}, $puzzle->{h});
    my $first = $puzzle->{first};
    my $shrinkable_height = 0;
    if ($OPT and $h > 2 and substr($first, $w + 1, 1) ne '='
            and substr($first, $w * 2 - 2) ne '=') {
        $shrinkable_height = 1;
    }
    my $shrinkable_width = 0;
    if ($OPT and $w > 2 and substr($first, $w + 1, 1) ne '='
            and substr($first, $w * ($h - 2) + 1) ne '=') {
        $shrinkable_width = 1;
    }
    my $head_row = substr($puzzle->{final}, 0, $puzzle->{w});
    my $head_col = &get_head_col($w, $h, $puzzle->{final});

    while (scalar(@{$puzzle->{queue}})) {
        if ((++$count % 100000) == 0) {
            printf "  iterate %d\n", $count;
        }
        if ($count > 1500000) {
            print "  OVER_ITER\n";
            last;
        }
        my $curr = shift @{$puzzle->{queue}};
        my $hist = $puzzle->{hash}->{$curr};
        my $movable = &get_movable($puzzle, $curr);
        foreach my $d (@$movable) {
            my $next = &apply_move2($moval, $curr, $d);
            if ($next eq $puzzle->{final}) {
                return $hist.$d;
            }
            # Check height shrinkable.
            if ($shrinkable_height and substr($next, 0, $w) eq $head_row) {
                printf "  Shrink height: %s\n", $hist.$d;
                my $puzzle2 = &new_puzzle($w, $h - 1, substr($next, $w));
                my $answer2 = &solve_puzzle($puzzle2);
                undef $puzzle2;
                if (defined $answer2) {
                    return $hist.$d.$answer2;
                }
            }
            # Check width shrinkable.
            if ($shrinkable_width
                    and &get_head_col($w, $h, $next) eq $head_col)
            {
                printf "  Shrink width: %s\n", $hist.$d;
                my $puzzle3 = &new_puzzle($w - 1, $h,
                    &remove_head_col($next, $w));
                my $answer3 = &solve_puzzle($puzzle3);
                undef $puzzle3;
                if (defined $answer3) {
                    return $hist.$d.$answer3;
                }
            }
            unless (exists $puzzle->{hash}->{$next}) {
                $puzzle->{hash}->{$next} = $hist.$d;
                push @{$puzzle->{queue}}, $next;
            }
        }
    }
    return undef;
}

sub apply_move2 {
    my ($moval, $s, $d) = @_;

    my @state = split //, $s;
    my $now = index $s, '0';
    my $next = $now;
    if (exists $moval->{$d}) {
        $next += $moval->{$d};
        @state[$now, $next] = @state[$next, $now];
    }

    return join('', @state);
}

sub get_movable {
    my ($puzzle, $curr) = @_;
    my $retval = [];

    my $pos = index $curr, '0';
    my $w = $puzzle->{w};
    my $x = $pos % $w;
    my $y = int($pos / $w);
    my @s = split //, $curr;

    if ($x > 0 and $s[$pos - 1] ne '=') {
        push @$retval, 'L';
    }
    if ($x < ($puzzle->{w} - 1) and $s[$pos + 1] ne '=') {
        push @$retval, 'R';
    }

    if ($y > 0 and $s[$pos - $w] ne '=') {
        push @$retval, 'U';
    }
    if ($y < ($puzzle->{h} - 1) and $s[$pos + $w] ne '=') {
        push @$retval, 'D';
    }

    return $retval;
}

sub get_final {
    my ($first) = @_;

    (my $numbers = $first) =~ s/[=0]//g;
    my @numbers = sort split //, $numbers;
    push @numbers, '0';

    my @final;

    (my $walls = $first) =~ s/[^=]/./g;
    for my $ch (split(//, $walls)) {
        if ($ch eq '.') {
            $ch = shift @numbers;
        }
        push @final, $ch;
    }

    return join('', @final);
}

sub get_head_col {
    my ($w, $h, $s) = @_;
    my $retval;
    for (my $i = 0; $i < length($s); $i += $w) {
        $retval .= substr($s, $i, 1);
    }
    return $retval;
}

sub remove_head_col {
    my ($s, $w) = @_;
    my $retval;
    for (my $i = 1; $i < length($s); $i += $w) {
        $retval .= substr($s, $i, $w - 1);
    }
    return $retval;
}

File: SlidePuzzle/solve4-1.cpp
#include <algorithm>
#include <cstdio>
#include <deque>
#include <map>
#include <string>
#include <vector>
#include <sstream>

using std::string;
using std::deque;
using std::map;
using std::vector;
using std::istringstream;

    string
get_final_state(const string& s)
{
    string s2;
    for (string::const_iterator i = s.begin(); i != s.end(); ++i)
        if (*i != '=' && *i != '0')
            s2 += *i;
    std::sort(s2.begin(), s2.end());
    s2 += '0';

    string s3;
    string::const_iterator r = s2.begin();
    for (string::const_iterator i = s.begin(); i != s.end(); ++i)
    {
        if (*i != '=')
            s3 += *r++;
        else
            s3 += '=';
    }

    return s3;
}

    int
get_pos(const string& s)
{
    return s.find_first_of('0');
}

    void
get_movable(vector<char>& movable, int w, int h, const string& s)
{
    int pos = get_pos(s);

    int x = pos % w;
    if (x > 0 && s[pos - 1] != '=')
        movable.push_back('L');
    if (x < (w - 1) && s[pos + 1] != '=')
        movable.push_back('R');

    int y = pos / w;
    if (y > 0 && s[pos - w] != '=')
        movable.push_back('U');
    if (y < (h - 1) && s[pos + w] != '=')
        movable.push_back('D');
}

    string
apply_move(const string& s, int w, char dir)
{
    int diff = 0;
    switch (dir)
    {
        case 'L': diff = -1; break;
        case 'R': diff = 1; break;
        case 'U': diff = -w; break;
        case 'D': diff = w; break;
    }

    string new_s = s;
    if (diff != 0)
    {
        int pos = get_pos(s);
        int new_pos = pos + diff;
        new_s[pos] = s[new_pos];
        new_s[new_pos] = '0';
    }

    return new_s;
}

    string
solve_puzzle2(int w, int h, const string& s)
{
    string retval;

    deque<string> queue;
    queue.push_back(s);
    map<string,string> hash;
    hash[s] = string("");
    string final = get_final_state(s);

    int count = 0;
    while (!queue.empty())
    {
        const string curr = queue.front();
        queue.pop_front();
        const string& hist = hash[curr];

        vector<char> movable;
        get_movable(movable, w, h, curr);
        for (vector<char>::const_iterator i = movable.begin();
                i != movable.end(); ++i)
        {
            string next = apply_move(curr, w, *i);
            string new_hist = hist;
            new_hist += *i;
            if (next == final) {
                return new_hist;
            }
            if (hash.find(next) == hash.end())
            {
                hash[next] = new_hist;
                queue.push_back(next);
            }
        }

        if ((++count % 500000) == 0) {
            printf("  iterate %d\n", count);
        }
    }

    return retval;
}

    string
solve_puzzle(int w, int h, const string& s)
{
    printf("Solving: %s\n", s.c_str());
    return solve_puzzle2(w, h, s);
}

    int
main(int argc, char** argv)
{
    for (++argv; *argv; ++argv)
    {
        int w, h;
        char ch;
        string puzzle;

        istringstream is(*argv);
        is >> w >> ch >> h >> ch >> puzzle;

        printf("(%d, %d): %s\n", w, h, puzzle.c_str());
        string answer = solve_puzzle(w, h, puzzle);
        printf("-> %s\n", answer.c_str());
    }

    return 0;
}
